squad/frontend/tests.py

```diff
@@ -46,18 +46,31 @@ def __init__(self):
         self.paginator = self
         self.paginator.num_pages = 0
         self.number = 0
+        self.all_tests = []
+
+    def __get_all_tests__(self, build, search):
+        self.all_tests = []
+
+        test_runs = TestRun.objects.filter(build=build).values('id')
+        test_runs_ids = [test_run['id'] for test_run in test_runs]
+        for chunk in split_list(test_runs_ids, chunk_size=100):
+            query_set = Test.objects.filter(test_run_id__in=chunk).prefetch_related('suite')
+            if search:
+                query_set = query_set.filter(name__icontains=search)
+
+            tests = query_set.only('id', 'suite', 'name', 'result', 'has_known_issues').order_by()
+            self.all_tests += tests
```

Comment:
this looks to me like you are loading all the tests, and then using the potentially huge list to calculate the pagination. the point of pagination is not having to load everything from the database in the first place.

Reply:
@terceiro previous code was running 3 significantly heavy queries: the first query did string matching + distinct() + count() just to get the count of tests. The second query (again with string matching) was ordering tests by number of failures/xfails/skips/passes, the result of this second query was just to build a dynamic [filter](https://github.com/Linaro/squad/blob/master/squad/frontend/tests.py#L119) for the third query that generates content for the test results page, which used suite_id and test name to query tests. Although test name is indexed, this would still take some time process.

All that was replaced by 2 much lighter queries: first one that does string matching and return all tests for a build. All previous calculations are made in memory: pagination, ordering and selection of the tests to be displayed. The second query simply passes the wanted tests ids.

Running these changes locally for an android-lkft project, it lowered loading times from 4000ms to just a couple of 100ms. Also, we are already doing something similar for workers, when loading tests to generate regressions and fixes.